sources/zarr/large_image_source_zarr/__init__.py

```diff
@@ -438,6 +488,204 @@ def getTile(self, x, y, z, pilImageAllowed=False, numpyAllowed=False, **kwargs):
+        """
+        Crop only applies to the output file, not the internal data access.
+
+        It consists of x, y, w, h in pixels.
+        """
+        return getattr(self, '_crop', None)
+
+    @crop.setter
+    def crop(self, value):
+        self._checkEditable()
+        if value is None:
+            self._crop = None
+            return
+        x, y, w, h = value
+        x = int(x)
+        y = int(y)
+        w = int(w)
+        h = int(h)
+        if x < 0 or y < 0 or w <= 0 or h <= 0:
+            msg = 'Crop must have non-negative x, y and positive w, h'
+            raise TileSourceError(msg)
+        self._crop = (x, y, w, h)
+
+    def write(
+        self,
+        path,
+        lossy=True,
+        alpha=True,
+        overwriteAllowed=True,
+    ):
+        """
+        Output the current image to a file.
+
+        :param path: output path.
+        :param lossy: if false, emit a lossless file.
+        :param alpha: True if an alpha channel is allowed.
+        :param overwriteAllowed: if False, raise an exception if the output
+            path exists.
+        """
+        # TODO: compute half, quarter, etc. resolutions
+        if not overwriteAllowed and os.path.exists(path):
+            raise TileSourceError('Output path exists (%s).' % str(path))
+
+        self._validateZarr()
+        suffix = Path(path).suffix
+        data_file = self._tempfile
+        data_store = self._zarr_store
+
+        if self.crop:
+            x, y, w, h = self.crop
+            current_arrays = dict(self._zarr.arrays())
+            # create new temp storage for cropped data
+            data_file = tempfile.NamedTemporaryFile()
+            data_store = zarr.SQLiteStore(data_file.name)
+            cropped_zarr = zarr.open(data_store, mode='w')
+            for arr_name in current_arrays:
+                arr = np.array(current_arrays[arr_name])
+                cropped_arr = arr.take(
+                    indices=range(x, x + w),
+                    axis=self._axes.get('x'),
+                ).take(
+                    indices=range(y, y + h),
+                    axis=self._axes.get('y'),
+                )
+                cropped_zarr.create_dataset(arr_name, data=cropped_arr, overwrite=True)
+                cropped_zarr.attrs.update(self._zarr.attrs)
+
+        data_file.flush()
+
+        if suffix in ['.db', '.sqlite']:
+            shutil.copy2(data_file.name, path)
+
+        elif suffix == '.zip':
+            zip_store = zarr.storage.ZipStore(path)
+            zarr.copy_store(data_store, zip_store)
+            zip_store.close()
+
+        elif suffix == '.zarr':
+            dir_store = zarr.storage.DirectoryStore(path)
+            zarr.copy_store(data_store, dir_store)
+            dir_store.close()
+
+        else:
+            from large_image_converter import convert
+
+            convert(data_file.name, path, overwrite=overwriteAllowed)
+
+        if self.crop:
+            data_file.close()
```

Comment:
I can imagine a scenario: create a sink, add some data, save, add some more data, save again.  With this `close`, does that work using crop?